src/framework/ui/uiwidget.cpp

```diff
@@ -606,6 +615,24 @@ void UIWidget::unlockChild(const UIWidgetPtr& child)
     }
 }
 
+void UIWidget::onAdopted(const UIWidgetPtr& parent)
+{
+    callLuaField("onAdopted", parent);
+
+    for (const UIWidgetPtr& child : m_children) {
+        child->onAdopted(static_self_cast<UIWidget>());
+    }
+}
+
+void UIWidget::onAbandoned(const UIWidgetPtr& parent)
+{
+    callLuaField("onAbandoned", parent);
+
+    for (const UIWidgetPtr& child : m_children) {
+        child->onAbandoned(static_self_cast<UIWidget>());
+    }
+}
+
```

Comment:
Same here, its going to trigger a lot of lua calls, we should avoid this at all cost.